Adopt smart pointers in RemoteWebInspectorUI.cpp

jelee53 opened the pull request:
#### 8f89f93f0265df983e34d1cd66b943f8c1d81aaa
<pre>
Adopt smart pointers in RemoteWebInspectorUI.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=295068">https://bugs.webkit.org/show_bug.cgi?id=295068</a>
<a href="https://rdar.apple.com/problem/154521813">rdar://problem/154521813</a>

Reviewed by Per Arne Vollan, Ryosuke Niwa, and Anne van Kesteren.

Smart pointer adoption as per the static analyzer.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::checkedDiagnosticLoggingClient const):
* Source/WebCore/page/Page.h:
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.cpp:
(WebKit::RemoteWebInspectorUI::RemoteWebInspectorUI):
(WebKit::RemoteWebInspectorUI::initialize):
(WebKit::RemoteWebInspectorUI::sendMessageToBackend):
(WebKit::RemoteWebInspectorUI::windowObjectCleared):
(WebKit::RemoteWebInspectorUI::frontendLoaded):
(WebKit::RemoteWebInspectorUI::changeSheetRect):
(WebKit::RemoteWebInspectorUI::setForcedAppearance):
(WebKit::RemoteWebInspectorUI::startWindowDrag):
(WebKit::RemoteWebInspectorUI::bringToFront):
(WebKit::RemoteWebInspectorUI::closeWindow):
(WebKit::RemoteWebInspectorUI::reopen):
(WebKit::RemoteWebInspectorUI::resetState):
(WebKit::RemoteWebInspectorUI::openURLExternally):
(WebKit::RemoteWebInspectorUI::revealFileExternally):
(WebKit::RemoteWebInspectorUI::save):
(WebKit::RemoteWebInspectorUI::load):
(WebKit::RemoteWebInspectorUI::pickColorFromScreen):
(WebKit::RemoteWebInspectorUI::showCertificate):
(WebKit::RemoteWebInspectorUI::setInspectorPageDeveloperExtrasEnabled):
(WebKit::RemoteWebInspectorUI::logDiagnosticEvent):
(WebKit::RemoteWebInspectorUI::protectedWebPage):
* Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.h:

Canonical link: <a href="https://commits.webkit.org/296884@main">https://commits.webkit.org/296884@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/723cfc80f4e1d924be6f10f8dc2180c7a9730c18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109802 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29460 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19890 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115823 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60039 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111765 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30138 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38048 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83446 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112750 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24021 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98880 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63909 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23400 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17028 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59617 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93391 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118615 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36841 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27294 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92449 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37214 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95147 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92272 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37239 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14983 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32680 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17731 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36736 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42206 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36396 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39738 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38105 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1591,6 +1591,11 @@ DiagnosticLoggingClient& Page::diagnosticLoggingClient() const
     return *m_diagnosticLoggingClient;
 }
 
+CheckedRef<DiagnosticLoggingClient> Page::protectedDiagnosticLoggingClient() const
```
I think we can call this member function 'checkedDiagnosticLoggingClient()'

pvollan commented on Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.cpp:
```diff
@@ -72,7 +72,7 @@ void RemoteWebInspectorUI::initialize(DebuggableInfoData&& debuggableInfo, const
     m_backendCommandsURL = backendCommandsURL;
 
-    m_page->protectedCorePage()->inspectorController().setInspectorFrontendClient(this);
+    WeakRef { m_page }->protectedCorePage()->inspectorController().setInspectorFrontendClient(this);
```
Perhaps we can create a helper method: `Ref<WebPage> RemoteWebInspectorUI::protectedWebPage()`?

pvollan commented on Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.cpp:
```diff
@@ -90,16 +90,17 @@ void RemoteWebInspectorUI::sendMessageToFrontend(const String& message)
 
-    m_frontendHost = InspectorFrontendHost::create(this, m_page->protectedCorePage().get());
-    m_frontendHost->addSelfToGlobalObjectInWorld(mainThreadNormalWorldSingleton());
+    frontendHost = InspectorFrontendHost::create(this, protectedWebPage()->protectedCorePage().get());
```
We should still assign this to `m_frontendHost`.

jelee53 replied:
Ugh that's so careless of me! Fixing.

rniwa commented on Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.cpp:
```diff
@@ -90,16 +90,16 @@ void RemoteWebInspectorUI::sendMessageToFrontend(const String& message)
 void RemoteWebInspectorUI::sendMessageToBackend(const String& message)
 {
-    WebProcess::singleton().parentProcessConnection()->send(Messages::RemoteWebInspectorUIProxy::SendMessageToBackend(message), m_page->identifier());
+    WebProcess::singleton().protectedParentProcessConnection()->send(Messages::RemoteWebInspectorUIProxy::SendMessageToBackend(message), m_page->identifier());
```
This is a bit silly. parentProcessConnection is a singleton just like `WebProcess::singleton()`. Maybe we can wrap in a helper function by the name of `parentProcessConnectionSingleton()`.

jelee53 replied:
I see. There's a lot of usages of protectedParentProcessConnection() outside of this file. Should I file a bugzilla ticket to replace all of those?

jelee53 replied:
Talked offline. We will be leaving this as-is.

cdumez commented on Source/WebKit/UIProcess/Cocoa/DiagnosticLoggingClient.h:
```diff
@@ -35,8 +35,9 @@
 namespace WebKit {
 
-class DiagnosticLoggingClient final : public API::DiagnosticLoggingClient {
+class DiagnosticLoggingClient final : public API::DiagnosticLoggingClient, public CanMakeCheckedPtr<DiagnosticLoggingClient> {
```
Can you explain this change? Where is this used in this PR? I looked at the PR twice but couldn't see any code using a CheckedPtr to a `WebKit::DiagnosticLoggingClient`, only to the WebCore one.

jelee53 replied:
I needed DiagnosticLoggingClient to become a CheckedPtr in order to make this helper function: 
```
Source/WebCore/page/Page.h:
WEBCORE_EXPORT CheckedRef<DiagnosticLoggingClient> checkedDiagnosticLoggingClient() const;`
```
which I eventually used in 
```
Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.cpp:
void RemoteWebInspectorUI::logDiagnosticEvent(const String& eventName,  const DiagnosticLoggingClient::ValueDictionary& dictionary)
{ 
.. 
protectedWebPage()->protectedCorePage()->checkedDiagnosticLoggingClient()->logDiagnosticMessageWithValueDictionary(eventName, "Remote Web Inspector Frontend Diagnostics"_s, dictionary, ShouldSample::No);
}
```

jelee53 replied:
Also, we will need `checkedDiagnosticLoggingClient()` to fix other unsafe cpp errors as well, such as with WebResourceLoader.cpp and WebInspectorUI.cpp: 

<img width="1272" alt="Screenshot 2025-07-01 at 12 14 03 AM" src="https://github.com/user-attachments/assets/d89878ea-9e51-4fc8-99f9-b60ccd82b739" />
<img width="908" alt="Screenshot 2025-07-01 at 12 14 43 AM" src="https://github.com/user-attachments/assets/c7371c59-6642-446d-8bd0-c7faba6f0b62" />

annevk replied:
Chris is pointing out that we have both `Source/WebCore/page/DiagnosticLoggingClient.h` and `Source/WebKit/UIProcess/Cocoa/DiagnosticLoggingClient.h`. And that you're not using the latter in this patch. And the former already has `CanMakeCheckedPtr`.

jelee53 replied:
Ohhh, I didn't even realize that there were two separate DiagnosticLoggingClient headers... Now that I'm reading Chris's comment closer - he's talking about a Webcore vs a Webkit one.

Anyways, even I was using the wrong file, I should have double-checked if the parent class was ref'd or check'd. 

You two are right. I don't need this change at all since I'm using the Webcore one not the Webkit. I'll back this out.

cdumez commented on Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.h:
```diff
@@ -144,8 +144,9 @@ class RemoteWebInspectorUI final
+    const Ref<WebPage> protectedWebPage();
 
-    WeakRef<WebPage> m_page;
+    const WeakRef<WebPage> m_page;
```
Why this change?

jelee53 replied:
Oops. Don't need it as we discussed. I forgot about this and will back it out.

jelee53 committed:
Adopt smart pointers in RemoteWebInspectorUI.cpp
https://bugs.webkit.org/show_bug.cgi?id=295068
rdar://problem/154521813

Reviewed by Per Arne Vollan, Ryosuke Niwa, and Anne van Kesteren.

Smart pointer adoption as per the static analyzer.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::checkedDiagnosticLoggingClient const):
* Source/WebCore/page/Page.h:
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.cpp:
(WebKit::RemoteWebInspectorUI::RemoteWebInspectorUI):
(WebKit::RemoteWebInspectorUI::initialize):
(WebKit::RemoteWebInspectorUI::sendMessageToBackend):
(WebKit::RemoteWebInspectorUI::windowObjectCleared):
(WebKit::RemoteWebInspectorUI::frontendLoaded):
(WebKit::RemoteWebInspectorUI::changeSheetRect):
(WebKit::RemoteWebInspectorUI::setForcedAppearance):
(WebKit::RemoteWebInspectorUI::startWindowDrag):
(WebKit::RemoteWebInspectorUI::bringToFront):
(WebKit::RemoteWebInspectorUI::closeWindow):
(WebKit::RemoteWebInspectorUI::reopen):
(WebKit::RemoteWebInspectorUI::resetState):
(WebKit::RemoteWebInspectorUI::openURLExternally):
(WebKit::RemoteWebInspectorUI::revealFileExternally):
(WebKit::RemoteWebInspectorUI::save):
(WebKit::RemoteWebInspectorUI::load):
(WebKit::RemoteWebInspectorUI::pickColorFromScreen):
(WebKit::RemoteWebInspectorUI::showCertificate):
(WebKit::RemoteWebInspectorUI::setInspectorPageDeveloperExtrasEnabled):
(WebKit::RemoteWebInspectorUI::logDiagnosticEvent):
(WebKit::RemoteWebInspectorUI::protectedWebPage):
* Source/WebKit/WebProcess/Inspector/RemoteWebInspectorUI.h:

Canonical link: https://commits.webkit.org/296884@main
